libgadget/forcetree.c

```diff
@@ -254,88 +256,72 @@ int get_freenode(int * nnext, struct NodeCache *nc)
     return ninsert;
 }
 
-/* Parent is a node where the subnode we want to add a particle to is filled.
- * We add a new internal node at this subnode and try to add both the old and new particles to it.
+/* Add a particle to a node in a known empty location.
  * Parent is assumed to be locked.*/
-int
-modify_internal_node(int parent, int subnode, int p_child, int p_toplace,
-        const ForceTree tb, int *nnext, struct NodeCache *nc, double minlen, int *closepairs)
+static int
+modify_internal_node(int parent, int subnode, int p_toplace, const ForceTree tb)
 {
-    int ret = 0;
-    int ninsert;
-    int child_subnode, new_subnode;
-    const int too_small = tb.Nodes[parent].len < minlen;
-    /* Just insert the particle if we have an empty spot.
-     * If the node is already too small, do not split it,
-     * but instead prepend the particle to a linked list.*/
-    if(p_child == -1 || too_small) {
-        ninsert = parent;
-        child_subnode = subnode;
-        new_subnode = subnode;
-    }
+    tb.Father[p_toplace] = parent;
+    tb.Nodes[parent].u.s.suns[subnode] = p_toplace;
+    return 0;
+}
+
+
+/* Create a new layer of nodes beneath the current node, and place the particle.
+ * Must have node lock.*/
+static int
+create_new_node_layer(int parent, int p_toplace,
+        const ForceTree tb, int *nnext, struct NodeCache *nc)
+{
+    int i;
+    int oldsuns[NCHILD+1];
+    struct NODE *nprnt = &tb.Nodes[parent];
+
+    /* Copy the old particles and a new one into a temporary array*/
+    memcpy(oldsuns, nprnt->u.s.suns, NCHILD * sizeof(int));
+    oldsuns[NCHILD] = p_toplace;
```

Comment:
Reusing the last element of oldsuns doesn't smell nice, but maybe there is no better way?

Reply:
You are right - there is a better way!